### PR TITLE
strict permission check when user enables network or GPU.

### PR DESCRIPTION
**Pull request by Sasasu**
if GPU or network not enabled, empty roles list means any user can use
    this runtime.

    if GPU or network enable, empty roles list means no one can use this
    runtime.

### CODE REVIEW

File: src/plc_configuration.c [hunk 1]
```diff
@@ -282,6 +282,9 @@ static void parse_runtime_configuration(xmlNode *node) {
 									"changing <enable_network> to \"yes\" automaticly");
 							conf_entry->enableNetwork = true;
 						}
+
+						// network will a new attack surface. enable strict permission check
+						conf_entry->useUserControl = true;
```
Comment (beeender) [inline src/plc_configuration.c:287]: This is not actually a good place for setting the config. The `enableNetwork` could be `false` at line 269.

I propose to have a unified place to set this at the end of this function/loop like:

```
if (conf_entry->enableNetwork || not_empty(roles) || enable_gpu) {
    conf_entry-> useUserControl = true;
}
```

File: src/plc_configuration.c [hunk 2]
```diff
@@ -619,7 +629,10 @@ static void print_runtime_configurations() {
 				}
 			}
 		}
-		if (conf_entry->useUserControl && conf_entry->roles != NULL) {
+		if (conf_entry->useUserControl) {
+			plc_elog(INFO, "    useUserControl  = 'true'");
```
Comment (beeender) [inline src/plc_configuration.c:633]: very minor. there are two spaces between `useUserControl` and `=`. So is line 636

File: src/plc_configuration.c [hunk 3]
```diff
@@ -848,15 +861,24 @@ char *get_sharing_options(runtimeConfEntry *conf, int container_slot, bool *has_
...
 
 bool plc_check_user_privilege(char *roles){
-
 	List *elemlist;
 	ListCell *l;
 	Oid currentUserOid;
 
+	if (roles == NULL) {
+		ereport(WARNING,
+				(errmsg("plcontainer: enable network access or physical device access need set a explicit user permission for the runtime"),
```
Comment (beeender) [inline src/plc_configuration.c:870]: To:

To access network or physical device, user permission needs to be granted explicitly in the runtime config.

File: tests/data/test_assign_gpu.py [hunk 4]
```diff
@@ -1,4 +1,4 @@
-#!/bin/env python2
+#!/usr/bin/env python2
```
Comment (beeender) [inline tests/data/test_assign_gpu.py:1]: hmmm, `/bin/env` has problems on some distros? which distro?

Comment (Sasasu) [reply tests/data/test_assign_gpu.py:1]: ubuntu18.04

File: src/plc_configuration.c (hunk 1 above)
Comment (Sasasu) [reply src/plc_configuration.c]: I want the check-and-fix section to be closer to the read-and-set section, so it's easier to distinguish dependencies. It is also easier when we split this large function.

fix `useUserControl = true when enableNetwork = 'no'`

**Review by beeender — APPROVED** [review]

File: src/plc_configuration.c [hunk 5]
```diff
@@ -848,15 +863,24 @@ char *get_sharing_options(runtimeConfEntry *conf, int container_slot, bool *has_
...
 
 bool plc_check_user_privilege(char *roles){
-
 	List *elemlist;
 	ListCell *l;
 	Oid currentUserOid;
 
+	if (roles == NULL) {
+		ereport(WARNING,
+				(errmsg("plcontainer: To access network or physical device, user permission needs to be granted explicitly in the runtime config."),
```
[the remark this answers is not in the document]
Comment (Sasasu) [reply src/plc_configuration.c:872]: I prefer not to use comment. need refactor the code struct.

split the very big config read logic `parse_runtime_configuration` and move the check-and-fix section to be closer to the read-and-set section.

need a refactor, not in this PR.

**Pull request merged**